### ISSUE TITLE
ARM64 - mount as directory does not work

### ISSUE THREAD

**Issue by billziss-gh** [winfsp/winfsp#448]
### Discussed in https://github.com/winfsp/winfsp/discussions/447

<div type='discussions-op-text'>

<sup>Originally posted by **overheadhunter** September 21, 2022</sup>
Another day, another question:

We encountered some problems when trying to mount to a path (as opposed to drive letter). Using memfs shows this:

```
> memfs-a64.exe -m C:\MOUNT
cannot mount MEMFS
The service memfs has failed to start (Status=c000000d).
```

However, `memfs-a64.exe -m M:` works just fine.

Is this a known issue on ARM? Can it be solved or is it a limitation we have to live with?</div>

**Comment by billziss-gh** [comment]
@overheadhunter

I believe I have identified the issue behind this problem. It is indeed a difference in how a fundamental Windows API behaves under x64 and ARM64.

This [`CreateFileW`](https://github.com/winfsp/winfsp/blob/v1.11/src/dll/mount.c#L633-L640) call in [`FspMountSet_Directory`](https://github.com/winfsp/winfsp/blob/v1.11/src/dll/mount.c#L602) succeeds in an x64 (and x86) process but fails with `ERROR_INVALID_PARAMETER` in an ARM64 process. The purpose of this call is to create the directory mountpoint.

The [`CreateFileW`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew) API is a fundamental Windows API for opening files and directories. It is also used to create new files and although this is not well documented it can be used for creating directories. This requires the specification of the flags `FILE_FLAG_BACKUP_SEMANTICS` and `FILE_FLAG_POSIX_SEMANTICS` and the file attribute `FILE_ATTRIBUTE_DIRECTORY`.

WinFsp uses this method to create directories when it also needs a handle to the just created directory. The alternative of using [`CreateDirectoryW`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createdirectoryw) to create a directory does not work in these cases, because it creates the directory but does not return a handle to the newly created directory. The alternative of using `CreateDirectoryW` to create the directory and then opening the directory using `CreateFileW`, unfortunately does not provide the semantics that WinFsp needs.

The ability of `CreateFileW` to create directories has been present in all versions of Windows I have tested on, except of course ARM64 Windows and only when running an ARM64 process. (It still works for x64 and x86 processes even on ARM64 Windows.) I am inclined to think of this as a Microsoft oversight or bug, because the behavior of a fundamental Windows API should not diverge across versions of Windows.

Nevertheless this creates a problem for us. While mounting on directories WinFsp requires semantics where a directory is created and the corresponding handle remains open and this can not be achieved in an ARM64 process. Luckily we can use the equivalent `NtCreateFile` to create the directory instead.

I am hoping to have a fix for this problem in the next release (v1.12B2).
